### ISSUE TITLE
Rails 5.2.0 - When Oj integrates with Rails, it modifies JSON and breaks Ohm

### ISSUE THREAD

**Issue by rgaufman** [soveran/ohm#233]
Hi there,

I have this code:

```ruby
require 'ohm'
require 'ohm/contrib'
Ohm.redis = Redic.new 'redis://127.0.0.1:6379/0'
class User < Ohm::Model
  attribute :user_id
  index :user_id
end
p User.find(user_id: 5)[0]
```

When I run it in plain pry, it works correctly.

```ruby
[8] pry(main)> p User.find(user_id: 5)[0]
nil
=> nil
```

However, when I run it in rails c, I get:
```ruby
[10] pry(main)> User.find(user_id: 5)[0]
RuntimeError: ERR Error running script (call to f_4bd605bfee5f1e809089c5f98d10fab8aec38bd3): @user_script:75: @user_script: 75: Lua redis() command arguments must be strings or integers
from /usr/local/lib/ruby/gems/2.5.0/gems/redic-1.5.0/lib/redic.rb:44:in `call!'
```

I modified /usr/local/lib/ruby/gems/2.5.0/gems/redic-1.5.0/lib/redic.rb to log the args being passed and in the case of plain Pry, it's doing this:
```ruby
["EVALSHA", "4bd605bfee5f1e809089c5f98d10fab8aec38bd3", 0, "[\"SISMEMBER\",\"User:indices:user_id:5\",0]"]
```

In the case of running from Rails c, it's doing this:
```ruby
["EVALSHA", "4bd605bfee5f1e809089c5f98d10fab8aec38bd3", 0, "[\"SISMEMBER\",{\"ns\":\"User:indices:user_id:5\",\"rc\":{\"url\":\"redis://127.0.0.1:6379/0\",\"client\":{\"semaphore\":{},\"connection\":{},\"uri\":\"redis://127.0.0.1:6379/0\",\"timeout\":10000000},\"buffer\":{}}},0]"]
```

The problem started with Rails 5.2.0 - I'm also running hiredis (0.6.1), redis (4.0.1), ohm (3.1.1), ohm-contrib from cyx/ohm-contrib.git (I tried to disable ohm-contrib, there was no change).

Any ideas why it is so different when running without rails and how to make it work?

**Comment by soveran** [comment]
Can you connect to the Redis server (maybe with redis-cli) and run the `MONITOR` command? That way we can see what Redis is receiving.

**Comment by rgaufman** [comment]
Sure thing, this is what Redis receives:

```
127.0.0.1:6379> MONITOR
OK
1526911240.655483 [0 127.0.0.1:55615] "EVALSHA" "4bd605bfee5f1e809089c5f98d10fab8aec38bd3" "0" "[\"SISMEMBER\",{\"ns\":\"User:indices:user_id:5\",\"rc\":{\"url\":\"redis://127.0.0.1:6379/0\",\"client\":{\"semaphore\":{},\"connection\":{},\"uri\":\"redis://127.0.0.1:6379/0\",\"timeout\":10000000},\"buffer\":{}}},0]"
```

**Comment by soveran** [comment]
OK, and can you do the same without Rails?

**Comment by rgaufman** [comment]
Without Rails:

```
$ redis-cli
127.0.0.1:6379> MONITOR
OK
1526911986.052211 [0 127.0.0.1:55802] "SELECT" "0"
1526911986.052295 [0 127.0.0.1:55802] "EVALSHA" "4bd605bfee5f1e809089c5f98d10fab8aec38bd3" "0" "[\"SISMEMBER\",\"User:indices:user_id:5\",0]"
1526911986.052322 [0 lua] "SISMEMBER" "User:indices:user_id:5" "0"
```

**Comment by soveran** [comment]
OK, thanks. My first guess is that Rails is modifying `String` or something from Ruby core that alters the behavior of `Nest`. I don't use Rails, but I will try install `ActiveSupport` and see if it alters the way `Ohm` works.

**Comment by soveran** [comment]
Ah, good that you tried that. I just tried something similar but I did `require "active_support"`, I wasn't aware of the `/all`. I also tried to install Rails but I couldn't do it because it can't compile `nokogiri`. Is there a way to install Rails without `nokogiri`?

**Comment by soveran** [comment]
Forget my last message, I was able to install Nokogiri and then Rails. I'll try to reproduce the error now.

**Comment by rgaufman** [comment]
Thank you :) 

I'm looking more into this as well, I noticed that this works in Rails:

```ruby
# In Rails
> User.all.entries
=> [#<User:0x00007fa83d6f4ec8 @_memo={}, @attributes={:user_id=>"asdasdasd"}, @id="1">]

# In Redis Monitor
1526915092.441547 [0 127.0.0.1:55986] "SMEMBERS" "User:all"
1526915092.441793 [0 127.0.0.1:55988] "HGETALL" "User:1"
```

So I wonder why specifically looking for first or using [0] doesn't:

```
# Rails
> User.all.first
RuntimeError: ERR Error running script (call to f_4bd605bfee5f1e809089c5f98d10fab8aec38bd3): @user_script:75: @user_script: 75: Lua redis() command arguments must be strings or integers
from /usr/local/lib/ruby/gems/2.5.0/gems/redic-1.5.0/lib/redic.rb:44:in `call!'
> User.all[0]
RuntimeError: ERR Error running script (call to f_4bd605bfee5f1e809089c5f98d10fab8aec38bd3): @user_script:75: @user_script: 75: Lua redis() command arguments must be strings or integers
from /usr/local/lib/ruby/gems/2.5.0/gems/redic-1.5.0/lib/redic.rb:44:in `call!'

# In Redis Monitor
1526915242.796220 [0 127.0.0.1:55988] "EVALSHA" "4bd605bfee5f1e809089c5f98d10fab8aec38bd3" "0" "[\"SORT\",{\"ns\":\"User:all\",\"rc\":{\"url\":\"redis://127.0.0.1:6379/0\",\"client\":{\"semaphore\":{},\"connection\":{},\"uri\":\"redis://127.0.0.1:6379/0\",\"timeout\":10000000},\"buffer\":{}}},\"LIMIT\",0,1]"
1526915188.323139 [0 127.0.0.1:55988] "EVALSHA" "4bd605bfee5f1e809089c5f98d10fab8aec38bd3" "0" "[\"SISMEMBER\",{\"ns\":\"User:all\",\"rc\":{\"url\":\"redis://127.0.0.1:6379/0\",\"client\":{\"semaphore\":{},\"connection\":{},\"uri\":\"redis://127.0.0.1:6379/0\",\"timeout\":10000000},\"buffer\":{}}},0]"
```

**Comment by soveran** [comment]
I have a file where I put `require "rails"`, but I'm not sure it's having any effect. What would you suggest as the minimal code that will allow me to reproduce the error? Should I create a Rails app?

**Comment by rgaufman** [comment]
That is a good idea! - I just tried to create a blank rails app, added ohm to it and it seems to work. Let me narrow down specifically what is causing this.

**Comment by soveran** [comment]
Cool, I will wait a bit to see if you can narrow it down. Maybe you can copy the app and start removing things. Thanks a lot for working on this!

**Comment by rgaufman** [comment]
I found the culprit!! 

One of the gems had this in it:

``` Ruby
# Optimised Javascript
require 'oj'
Oj.optimize_rails
```

Which optimises the json parser, but I guess breaks Ohm!! - Any ideas how to make Ohm work with Oj? < https://github.com/ohler55/oj

**Comment by soveran** [comment]
Can you try with this patch?

```diff
diff --git a/lib/ohm.rb b/lib/ohm.rb
index 8cd5c1a..f6fbfcd 100644
--- a/lib/ohm.rb
+++ b/lib/ohm.rb
@@ -1353,10 +1353,10 @@ module Ohm
       end
 
       @id = script(LUA_SAVE, 0,
-        features.to_json,
-        _sanitized_attributes.to_json,
-        indices.to_json,
-        uniques.to_json
+        JSON.dump(features),
+        JSON.dump(_sanitized_attributes),
+        JSON.dump(indices),
+        JSON.dump(uniques)
       )
 
       return self
@@ -1377,13 +1377,16 @@ module Ohm
         uniques[field] = value.to_s
       end
 
+      features = {
+        "name" => model.name,
+        "id" => id,
+        "key" => key.to_s
+      }
+
       script(LUA_DELETE, 0,
-        { "name" => model.name,
-          "id" => id,
-          "key" => key.to_s
-        }.to_json,
-        uniques.to_json,
-        model.tracked.to_json
+        JSON.dump(features),
+        JSON.dump(uniques),
+        JSON.dump(model.tracked)
       )
 
       return self
```

I can also create a branch if that's easier for you. Let me know.

**Comment by rgaufman** [comment]
Hmm, strange, I applied the patch and I'm still getting:

```ruby
[2] pry(main)> User.find(user_id: 5)[0]
["EVALSHA", "4bd605bfee5f1e809089c5f98d10fab8aec38bd3", 0, "[\"SISMEMBER\",\"User:indices:user_id:5\",0]"]
=> nil
[3] pry(main)> Oj.optimize_rails
=> nil
[4] pry(main)> User.find(user_id: 5)[0]
["EVALSHA", "4bd605bfee5f1e809089c5f98d10fab8aec38bd3", 0, "[\"SISMEMBER\",{\"ns\":\"User:indices:user_id:5\",\"rc\":{\"url\":\"redis://127.0.0.1:6379\",\"client\":{\"semaphore\":{},\"connection\":{},\"uri\":\"redis://127.0.0.1:6379\",\"timeout\":10000000},\"buffer\":{}}},0]"]
RuntimeError: ERR Error running script (call to f_4bd605bfee5f1e809089c5f98d10fab8aec38bd3): @user_script:75: @user_script: 75: Lua redis() command arguments must be strings or integers
from /usr/local/lib/ruby/gems/2.5.0/gems/redic-1.5.0/lib/redic.rb:44:in `call!'
```

**Comment by soveran** [comment]
I created a Rails app and was able to reproduce the error. I have a fix already, but I'm afraid I will have to release a new version of Ohm. Before releasing it, I'll make sure I've got the right implementation.

**Comment by rgaufman** [comment]
Thank you very much for looking at this, please let me know if you want me to test anything :)

**Comment by soveran** [comment]
Here's an example of the issue we have:

```ruby
require "oj"
require "json"

class MyString < String
end

mystr = MyString.new("foo")

test "encoding instances of String" do
  assert_equal "\"foo\"", JSON.dump("foo")
  assert_equal "\"foo\"", Oj.dump("foo")
end

test "encoding instances of MyString" do
  assert_equal "\"foo\"", JSON.dump(mystr)
  assert_equal "{\"^o\":\"MyString\",\"self\":\"foo\"}", Oj.dump(mystr)
end

Oj.mimic_JSON

test "encoding instances of MyString with mimic_JSON" do
  assert_equal "\"foo\"", JSON.dump(mystr)
  assert_equal "\"foo\"", Oj.dump(mystr)
end
```

When `Oj` has to dump an object that inherits from `String`, it behaves differently than what `JSON` does. That shouldn't be a problem, but sadly `Oj` modifies `JSON` when it's integrated with Rails. Note that when I called `Oj.mimic_JSON`, the behavior was corrected.

Something you can try in your code is to call `Oj.mimic_JSON` to see if it solves the problem. In the end, I will have to modify `Ohm` to prevent it from be affected by that bug in `Oj`.

**Comment by soveran** [comment]
There's something you can try, I think it can be the final workaround.

Modify `Nest` and add this method:

```ruby
def to_json(*args)
  @ns.to_json(*args)
end
```

If it works for you, I will release a new version of Nest with that change.

**Comment by rgaufman** [comment]
That works beautifully, thank you!!

**Comment by soveran** [comment]
The change was released with Nest 3.1.2. Thanks a lot!
